src/util/heap/layout/mod.rs

```diff
@@ -34,3 +34,37 @@ pub fn create_mmapper() -> Box<dyn Mmapper> {
     // TODO: ByteMapMmapper for 39-bit or less virtual space
     Box::new(fragmented_mapper::FragmentedMapper::new())
 }
+
+use crate::util::Address;
+
+/// Return true if the given address in our heap range
+pub fn address_in_heap(addr: Address) -> bool {
+    addr >= vm_layout_constants::HEAP_START && addr < vm_layout_constants::HEAP_END
+}
+
+/// Return true if the given address in our available heap range (where we manage and allocate)
+pub fn address_in_avialable_range(addr: Address) -> bool {
+    addr >= vm_layout_constants::AVAILABLE_START && addr < vm_layout_constants::AVAILABLE_END
+}
+
+/// Return true if the given range overlaps with our heap range
+pub fn range_overlaps_heap(addr: Address, size: usize) -> bool {
+    !(addr >= vm_layout_constants::HEAP_END || addr + size <= vm_layout_constants::HEAP_START)
+}
+
+/// Return true if the given range overlaps with our available heap range
+pub fn range_overlaps_available_range(addr: Address, size: usize) -> bool {
+    !(addr >= vm_layout_constants::AVAILABLE_END
+        || addr + size <= vm_layout_constants::AVAILABLE_START)
+}
```

Comment:
Instead of implementing methods for testing whether an address is within range, I suggest we use the `Range<Address>` type (i.e. `start..end`) and provide a few functions (or constants) that return ranges, for example,

```rust
pub fn heap_range() -> Range<Address> {
    vm_layout_constants::HEAP_START..vm_layout_constants::HEAP_END
}
```

Then we can use standard methods for testing range inclusion

```rust
if heap_range().contains(addr) { ... }
```

We can implement a method for computing the intersection of two ranges to make the code more readable:

```rust
impl Range<Address> {
    fn intersection(&self, other: &Range<Address>) -> Range<Address> {
        self.start.max(other.start)..self.end.min(other.min)
    }
}
```

(Note that there is a `range` crate that provides the same method (https://docs.rs/range/latest/range/struct.Range.html#method.intersect), but it seems to be unmaintained for quite a few years.)

Then we can test whether two ranges overlap by testing if their intersection is empty

```rust
let overlap = (start..end).intersect(available_range());
if !overlap.is_empty() {
    args.plan_args.vm_map.insert(overlap.start, overlap.len(), rtn.descriptor);
}

Reply:
Good point. I have made changes as suggested.